### PR TITLE
Admin & Mailer: created controllers, views, mailers, models, routes

### PR DESCRIPTION
**Pull request by curtydudes**
**Story:** 
This PR is for User Story 3 &4 and Admin Story 1-6* (1-3 are still on the way and in development)

The PR is quite big but the thought process was that in approving the users, it had to go side by side with the mailer but while the mailer is still receiving an error and can't send the mails to gmail, the admin approvals is working quite well. 

The idea here is to have an admin account that has several routes to an approval page and a creating user page while the homepage will house the admitted users. 

**References**
https://github.com/heartcombo/devise/wiki/How-To:-Require-admin-to-activate-account-before-sign_in
https://blog.saeloun.com/2019/10/10/rails-6-adds-support-for-multi-environment-credentials.html#how-to-read-credentials
https://dev.to/morinoko/sending-emails-in-rails-with-action-mailer-and-gmail-35g4

**Screenshots**
This is the admin home page where you can also see "Create User" where the Admin Story 1-3 will be in place. This is where the admin will see the accepted and approved users
image

This is the admin approval page. This is where the admin will approve the users. 
image

This is the result of approving a user after
image

ADD/CHANGES:
1) Admin Controller - I have add controllers for admitting the user to the site while also putting the add, view, edit user in place as well 
2) Application Controller - Since the admin should be routed to a different place and since the admin and user was generated by devise, I had to make an if statement when signing in. 
3) Added an index for Users Controller for my routing 
4) Generated an admin mailer via devise
5) Edited application mailer to send it to my dummy email
6) Approved and welcome mailer generated for the emailing of users
7) edited the user model as per the article from devise https://github.com/heartcombo/devise/wiki/How-To:-Require-admin-to-activate-account-before-sign_in
8) Created an approvals page
9) Created a homepage for admins via users-portfolio
10) Credentials: added gmail credentials 
11) Config changes based on  https://dev.to/morinoko/sending-emails-in-rails-with-action-mailer-and-gmail-35g4
12) added routes for admin functionalities
13) db migration based on the devise documentation and had to delete the previous column

### CODE REVIEW

**Comment by curtydudes** [comment]
Hi maam @analyn-cajocson, this is weird. The PR failed the test due to the Rubocop, but in my terminal it's all corrected. 
image

File: app/views/admins/add_user.html.erb [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+        <%= f.text_field :first_name, autofocus: true, autocomplete: :off, placeholder: "First Name" %>
+        <%= f.text_field :last_name, autofocus: true, autocomplete: :off, placeholder: "Last Name" %>
+        <%= f.text_field :username, autofocus: true, autocomplete: :off, placeholder: "Username" %>
+        <%= f.password_field :password, autocomplete: "new-password", placeholder: "Password (6 characters minimum)" %>
+        <%= f.password_field :password_confirmation, autocomplete: "new-password", placeholder: "Confirm Password" %>
+      </div>
+      <div>
+        <%= f.submit "Create Account" %>
+      </div>
+    <% end %>
```
Comment (analyn-cajocson) [inline app/views/admins/add_user.html.erb]: Fix code alignment: 2spaces
add new line at the end of file

File: app/views/admins/approvals.html.erb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  <% @users.each do |user| %>
+  <div>
+    <h5><%= user.id %></h5>
+    <h5><%= user.username %></h5>
+    <h5><%= user.email %></h5>
+    <%= link_to "Admit",  admins_approvals_path(id: user.id, admitted: true), method: :put %>
+  </div>
+  <% end %>
+  </div>
+</div>
```
Comment (analyn-cajocson) [inline app/views/admins/approvals.html.erb]: add new line

File: app/views/admins/approvals.html.erb [hunk 3]
```diff
@@ -0,0 +1,24 @@
+<%= csrf_meta_tags %>
```
Comment (analyn-cajocson) [inline app/views/admins/approvals.html.erb]: why do you have to add this? I think this is already in `application.html.erb`

File: app/views/admins/user_portfolio.html.erb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+            <td><%= user.email %>
+            <td><%= user.approved %>
+            <td><%= link_to "Edit", edit_user_path(user) %>
+          </tr>
+        <% end %>
+      </table>
+       <% end %>
+
+  </section>
+</section>
```
Comment (analyn-cajocson) [inline app/views/admins/user_portfolio.html.erb]: add new line

File: app/views/layouts/mailer.html.haml [hunk 5]
```diff
@@ -0,0 +1,3 @@
+%hmtl
+  %body
+    = yield
```
Comment (analyn-cajocson) [inline app/views/layouts/mailer.html.haml]: add new line

File: app/views/users/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<h2>User Details</h2>
+
+<table>
+    <tbody>
+      <tr>
+        <td><strong>Name: </strong> <%= @user.first_name %><%= @user.last_name %></td>
+        <td><strong>Email: </strong> <%= @user.email %></td>
+      </tr>
+    </tbody>
+  </table>
```
Comment (analyn-cajocson) [inline app/views/users/index.html.erb]: add new line

File: config/environments/production.rb [hunk 7]
```diff
@@ -109,4 +109,17 @@
...
+  config.action_mailer.delivery_method = :smtp
+  host = 'https://rails-stock-trading-app.herokuapp.com/' #replace with your own url
+  config.action_mailer.default_url_options = { host: host }
+
+  # SMTP settings for gmail
+  config.action_mailer.smtp_settings = {
+    :address              => "smtp.gmail.com",
+    :port                 => 587,
+    :user_name            => <gmail_username>,
+    :password             => <gmail_password>,
```
Comment (analyn-cajocson) [inline config/environments/production.rb]: how did you store the value of `<gmail_username>` and  `<gmail_password>`?

**Review by analyn-cajocson — CHANGES_REQUESTED** [review]

File: config/environments/production.rb (hunk 7 above)
Comment (curtydudes) [reply config/environments/production.rb]: hi maam @analyn-cajocson , this was stored in credentials with the thought process that it should be encrypted or hidden. Was this the right thing to do or is it advisable to have it written there in the production.rb itself?

Comment (analyn-cajocson) [reply config/environments/production.rb]: Hmm maybe you are not doing it right. Have you checked to `rails console` if you can access the `<gmail_username> `and `<gmail_password>`?

File: app/views/admins/approvals.html.erb (hunk 3 above)
Comment (curtydudes) [reply app/views/admins/approvals.html.erb]: hi maam, you're right. This was not needed and is already in application.html.erb. The intention there was that at some point, I was receiving an ActionController::InvalidAuthenticityToken error going to approvals.html.erb and the code was one of the solutions supposedly in[ stock overflow ](https://stackoverflow.com/questions/3364492/actioncontrollerinvalidauthenticitytoken). It happened to work when I bumped into the problem the first time in application.html.erb and I thought this was the solution the second time around. 

Will delete this line, maam. I did see that the local host is working well without the line. Thanks

File: config/environments/production.rb (hunk 7 above)
Comment (curtydudes) [reply config/environments/production.rb]: I think I am not doing it right. I followed [this site](https://dev.to/morinoko/sending-emails-in-rails-with-action-mailer-and-gmail-35g4) and it said "Replace <gmail_username> and <gmail_password> with your own username and password, which would preferably be hidden away as environment variables." I might have assumed to edit it in credentials. I'll have it written in the production.rb instead

Comment (analyn-cajocson) [reply config/environments/production.rb]: Refer to this: https://blog.saeloun.com/2019/10/10/rails-6-adds-support-for-multi-environment-credentials.html#how-to-read-credentials

Comment (curtydudes) [reply config/environments/production.rb]: Hi maam @analyn-cajocson , based on the first commit I am able to access the rails console with my gmail username and pw. Will this do?

image

Comment (analyn-cajocson) [reply config/environments/production.rb]: You should do that to access your credentials inside your app hehe

**Review by analyn-cajocson — APPROVED** [review]

**Pull request merged**